Code/GraphMol/Fingerprints/MHFP.cpp

```diff
@@ -0,0 +1,288 @@
+    uint32_t a = dist_a(rand);
+    uint32_t b = dist_b(rand);
+
+    while (std::find(perms_a_.begin(), perms_a_.end(), a) != perms_a_.end()) {
+      a = dist_a(rand);
+    }
+
+    while (std::find(perms_b_.begin(), perms_b_.end(), b) != perms_b_.end()) {
+      b = dist_a(rand);
+    }
+
+    perms_a_[i] = a;
+    perms_b_[i] = b;
+  }
+}
+
+std::vector<uint32_t>
+MHFPEncoder::FromStringArray(const std::vector<std::string>& vec)
+{
+  std::vector<uint32_t> mh(n_permutations_, max_hash_);
+
+  for (uint32_t i = 0; i < vec.size(); i++) {
+    auto hashi = FNV::hash(vec[i]);
+    for (size_t j = 0; j < n_permutations_; j++){
+      uint32_t tmp = (FastMod((perms_a_[j] * hashi + perms_b_[j]), prime_)) & max_hash_;
+      mh[j] = std::min(tmp, mh[j]);
+    }
+  }
+
+  return mh;
+}
+
+std::vector<uint32_t>
+MHFPEncoder::FromArray(const std::vector<uint32_t>& vec)
+{
+  std::vector<uint32_t> mh(n_permutations_, max_hash_);
+
+  for (uint32_t i = 0; i < vec.size(); i++) {
+    for (size_t j = 0; j < n_permutations_; j++){
+      uint32_t tmp = (FastMod((perms_a_[j] * vec[i] + perms_b_[j]), prime_)) & max_hash_;
+      mh[j] = std::min(tmp, mh[j]);
+    }
+  }
+
+  return mh;
+}
+
+std::vector<std::string>
+MHFPEncoder::CreateShingling(const ROMol& mol, 
+               const unsigned char& radius,
+               const bool& rings,
+               const bool& isomeric,
+               const bool& kekulize,
+               const unsigned char& min_radius) {
+  std::vector<std::string> shingling;
+
+  if (rings) {
+    VECT_INT_VECT bonds_vect = mol.getRingInfo()->bondRings();
+
+    for (size_t i = 0; i < bonds_vect.size(); i++) {
+      std::unique_ptr<ROMol> m(Subgraphs::pathToSubmol(mol, bonds_vect[i]));
+      shingling.emplace_back(MolToSmiles(*m));
+    }
+  }
+
+  unsigned char min_radius_internal = min_radius;
+
+  if (min_radius == 0) {
+    for (auto atom : mol.atoms()) {
+      shingling.emplace_back(SmilesWrite::GetAtomSmiles(atom, false, nullptr, false, true));
+    }
+
+    min_radius_internal++;
+  }
+
+  uint32_t index = 0;
+  for (auto atom : mol.atoms()) {
+    for (unsigned char r = min_radius_internal; r < radius + 1; r++) {
+      PATH_TYPE path = findAtomEnvironmentOfRadiusN(mol, r, index);
+      INT_MAP_INT amap;
+      std::unique_ptr<ROMol> submol(Subgraphs::pathToSubmol(mol, path, false, amap));
+
+      if (amap.find(index) == amap.end())
+        continue;
+
+      std::string smiles = MolToSmiles(*submol, isomeric, kekulize, amap[index]);
+      
+      if (smiles != "")
+        shingling.emplace_back(smiles);
+    }
+
+    index++;
+  }
+
+  return shingling;
+}
+
+std::vector<std::string>
+MHFPEncoder::CreateShingling(std::string& smiles, 
+               const unsigned char& radius,
+               const bool& rings,
+               const bool& isomeric,
+               const bool& kekulize,
+               const unsigned char& min_radius) {
+  std::unique_ptr<ROMol> m(SmilesToMol(smiles));
```

Comment:
To be super safe, you should make sure this pointer was actually created before you dereference it. One possibility:
```suggestion
  std::unique_ptr<ROMol> m(SmilesToMol(smiles));
  PRECONDITION(m,"could not parse smiles");
```
It would be better to either throw a `ValueErrorException()` or return an empty result, whichever you think makes the most sense.